src/ruby/puppet-server-lib/puppet/server/master.rb

```diff
@@ -57,6 +60,53 @@ def handleRequest(request)
         response["X-Puppet-Version"])
   end
 
+  def getClassInfoForAllEnvironments()
+    # The clear_environment_settings(env) ensures that the calls which happen
+    # later on to enumerate all of the manifests for each environment are
+    # using the latest environment.conf settings for each environment.  Without
+    # this call, cached (and, therefore, possibly stale) environment.conf
+    # settings would be used.  We want to return the latest data for any calls
+    # made to this method.
+    @env_loader.list.each do |env|
+      Puppet.settings.clear_environment_settings(env)
+    end
+
+    environments =
+      Hash[@env_loader.list.collect do |env|
+       [env.name, self.class.getManifests(env)]
+      end]
```

Comment:
I don't really care about the implementation details other than the following:

* I am terrified of the idea of you calling {{Puppet.settings.clear_environment_settings(env)}}, and to a lesser degree, even {{Puppet.lookup(environments)}}, from our code.  I would much prefer that we tell @hlindberg exactly what data we need to ask them for and he provides a method for us to call to get it, so that he is free to change implementation details w/o us needing to worry about it.  IMO we should not be mucking directly with anything related to Puppet settings, we should let that happen on the Puppet side of the API.
* In the grand and ideal world of the future, we would be able to parse the environment configs directly from Clojure and do the mapping from "list of changed files" to "list of affected environments" on our own, rather than having to call into Ruby for that.  So whatever we do now, it'd be nice for it to be encapsulated in some way that made it fairly easy to switch to that model if we're ever able to do so.

Beyond those constraints it doesn't matter much to me what the Ruby API ends up looking like as long as both you and @hlindberg are in agreement on it, so feel free to work it out with his team however you think it would be best.